x-pack/auditbeat/processors/sessionmd/processdb/reaper.go

```diff
@@ -60,50 +40,59 @@ func (h *rcHeap) Pop() any {
 // assumption will need to be revisited.
 func (db *DB) startReaper() {
 	go func(db *DB) {
-		ticker := time.NewTicker(reaperInterval)
+		ticker := time.NewTicker(db.reaperPeriod)
 		defer ticker.Stop()
 
-		h := &db.removalCandidates
-		heap.Init(h)
 		for {
 			select {
 			case <-ticker.C:
-				db.mutex.Lock()
-				now := time.Now()
-				for {
-					if len(db.removalCandidates) == 0 {
-						break
-					}
-					v := heap.Pop(h)
-					c, ok := v.(removalCandidate)
-					if !ok {
-						db.logger.Debugf("unexpected item in removal queue: \"%v\"", v)
-						continue
-					}
-					if now.Sub(c.exitTime) < removalTimeout {
-						// this candidate hasn't reached its timeout, put it back on the heap
-						// everything else will have a later exit time, so end this run
-						heap.Push(h, c)
-						break
-					}
-					p, ok := db.processes[c.pid]
-					if !ok {
-						db.logger.Debugf("pid %v was candidate for removal, but was already removed", c.pid)
-						continue
-					}
-					if p.PIDs.StartTimeNS != c.startTime {
-						// this could happen if the PID has already rolled over and reached this PID again.
-						db.logger.Debugf("start times of removal candidate %v differs, not removing (PID had been reused?)", c.pid)
-						continue
-					}
-					delete(db.processes, c.pid)
-					delete(db.entryLeaders, c.pid)
-					delete(db.entryLeaderRelationships, c.pid)
+				if !db.skipReaper {
+					db.reapProcs()
 				}
-				db.mutex.Unlock()
 			case <-db.stopChan:
 				return
 			}
 		}
 	}(db)
 }
+
+// run as a separate function to make testing easier
+func (db *DB) reapProcs() {
+	now := time.Now()
+	db.mutex.Lock()
+	db.logger.Debugf("REAPER: processes: %d removal candidates: %d", len(db.processes), len(db.removalMap))
+	for pid, cand := range db.removalMap {
+		if functionTimeoutReached(now, cand.exitTime) {
+			// this candidate hasn't reached its timeout
+			continue
+		}
+
+		p, ok := db.processes[pid]
+		if !ok {
+			// this represents an orphaned exit event with no matching exec event.
+			// in this case, give us a few iterations for us to get the exec, since things can arrive out of order.
+			if cand.removeAttempt < exitRemoveAttempts {
+				cand.removeAttempt += 1
+				db.removalMap[pid] = cand
```

Comment:
Is this needed? I don't see it being removed prior to this point.

Reply:
Not sure what you mean?

Reply:
Yes, that updates the existing entry in the map. The compiler won't let you do `map[key].struct_val = new`, if that's what you're thinking.

Reply:
I was thinking that line 75 changes what's in the map.